File: Diagonal_Traverse_Matrix.cpp
// Time Complexity : O(n*m)
// Space Complexity :O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this :No


// Your code here along with comments explaining your approach


class Solution {
public:
    vector<int> findDiagonalOrder(vector<vector<int>>& matrix) {
         int n = matrix.size();
        
        // return null if the input matrix is null
        if (n ==0){
            vector<int> result;
            return result;
        }
        int m = matrix[0].size();
      

        //initializing the variables
        int direction = 1;
        int iteration = 0;  
        int row = 0;
        int column = 0;
         vector<int> result(m*n, 0);
        // return null if the first row is null
        
    if (m ==0){
        return result;
    }
        
        while (iteration < (m)*(n)){
           
            result[iteration] = matrix[row][column]; // appending the value to the result array
            iteration +=1; // counter
           
            // code for moving up wards i.e. directio is 1
            if ( direction == 1){
               if ( column == m-1){
                   direction = -1;
                  row +=1;
               } 
                else if (row == 0) {
                    direction = -1;
                    column +=1;
                }
                else {
                    row -=1;
                    column +=1;
                }
            } else {
                
         // code for moving downwards i.e direction = -1
               if ( row == n-1){
                   direction = 1;
                  column +=1;
               } 
                else if (column == 0) {
                    direction = 1;
                    row +=1;
                }
                else {
                    row +=1;
                    column -=1;
                }
            }
            
            
            
            
            
        }
        
      return result;  
        
    }
};

File: diagonal_traverse_matrix.py
# Time Complexity : O(m*n)
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No
from typing import List


class Solution:
    def findDiagonalOrder(self, mat: List[List[int]]) -> List[int]:
        m = len(mat)
        n = len(mat[0])
        _sol = []
        _idx = 0
        _dir = 1
        col = 0
        row = 0
        while _idx < (n * m):
            _sol.append(mat[row][col])
            _idx += 1
            if _dir == 1:
                if col == n - 1:
                    _dir = -1
                    row += 1
                elif row == 0:
                    col += 1
                    _dir = -1
                else:
                    row -= 1
                    col += 1
            else:
                if row == m - 1:
                    _dir = 1
                    col += 1
                elif col == 0:
                    row += 1
                    _dir = 1
                else:
                    col -= 1
                    row += 1

        return _sol


if __name__ == '__main__':
    # _mat = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
    _mat = [[2, 3]]
    print(Solution().findDiagonalOrder(_mat))
